designs/2023-only-run-reporting-rules/README.md

```diff
@@ -0,0 +1,129 @@
+of overhead on ESLint runtime. On an example large codebase, I found that by disabling all
+rules set to warn, I was able to reduce ESLint runtime from 4 minutes and 39 seconds to just
+22 seconds. As the `warn` rules are entirely ignored in this situation, this is a substantial
+amount of wasted time and resources that could be avoided.
+
+## Detailed Design
+
+The proposed implementation is to add a `--skip-warnings` flag to ESLint that will skip
+running any rules that are set to the `warn` level. This will be a CLI flag only, as it
+doesn't make sense to have this flag in a configuration file. This flag would implicitly
+imply the `--quiet` flag, given if the `warn` rules are not run, they will not be outputted
+to reporters.
+
+From an API perspective, this would be implemented by a predicate function that acts as
+a filter on which rules should be run. The predicate would take the rule configuration,
+and return whether the rule should be run. For simplicity, this function should not be
+given rules marked as `off`, as if this function handled existing behavior then users of the
+API would have to mimic that when attempting to extend it.
+
+The flag would be implemented by passing a predicate function that only returns true for
+rules that have been set to `error`, thus filtering out any that are marked as `warn`.
+
+The check for unused `eslint-disable` directives (`--report-unused-disable-directives`)
+should continue to mark `warn` rules as used, even when running with `--skip-warnings`.
+
+The `--max-warnings` flag should disable the `--skip-warnings` flag, as it requires that
+`warn` rules are run to know the current number of reported warnings.
+
+## Documentation
+
+Both the flag and API would make sense to document. As they are relatively simple additions,
+they would be documented inline with the rest of the CLI and API documentation. Due to the
+potentially major performance improvements for many setups, I feel mentioning it in the
+release blog post, alongside the potential benefits, would be a good idea to spread awareness.
+
+## Drawbacks
+
+This change adds a new command line flag, as well as a new API method. This adds further
+maintenance burden. However, the implementation is relatively simple, and the benefits
+are significant.
+
+Another potential drawback around this specific implementation is that the existence of
+both `--quiet` and `--skip-warnings` might be confusing to some users, however this can be
+partially alleviated by updating the `--quiet` docs to mention that warnings will still be
+run, and to use `--skip-warnings` instead to not run them.
+
+## Backwards Compatibility Analysis
+
+The proposed solution does not cause any compatibility issues, as all functionality is
+implemented via new APIs and command line flags. Existing users will be unaffected
+by this change.
+
+## Alternatives
+
+### Implement this into `--quiet` instead of a new flag
```

Comment:
Why exactly would it be a breaking change to stop running rules that won't influence the results?

Reply:
I suppose that rules could be modifying shared state or have other side effects while running? I'm not sure if that's supposed to be allowed or not, but some plugins might be doing it.

Regardless of that, it sounds like there's another potential breaking change related to the interaction with `--max-warnings` flag.

Reply:
what if it only skipped the rules when `max-warnings` wasn't provided? Then it would automatically do what made the most sense.

Reply:
> I suppose that rules could be modifying shared state or have other side effects while running? I'm not sure if that's supposed to be allowed or not or if we need to worry about it here.

We have one case where rule API actually encourages side effects: [`context.markVariableAsUsed()`](https://github.com/eslint/eslint/blob/d8c8ede088e1f4a82c9b6b5c2772af268b9161aa/lib/linter/linter.js#L1031). For example, [`react/jsx-uses-vars`](https://github.com/jsx-eslint/eslint-plugin-react/blob/13d23b898de83557861b0e8bd419477f122e4839/lib/rules/jsx-uses-vars.js) calls this so that `no-unused-vars` doesn't report variables that are referenced in jsx elements. If someone has `"react/jsx-uses-vars": "warn"` in their config, and we don't run this rule, then they'll start getting `no-unused-vars` errors.

I'm also in favor of not introducing another flag, but not running rules set to "warn" indeed looks like a breaking change even without `--max-warnings`.

Reply:
Good point.

Reply:
> I'm also in favor of not introducing another flag, but not running rules set to "warn" indeed looks like a breaking change even without --max-warnings.

I'd be okay with breaking this use case because it's a super-edge case and we've always said that rules depending on what other rules are doing is not dependable. We can always give them a heads up ahead of time.

I think there's a lot more good to be done by not running rules that won't be reported vs. keeping the status quo for this edge case.

And I also like the idea of just implementing this behind `--quiet` rather than introducing a new flag.

Reply:
I would kindly disagree with implementing this under the --quiet flag. I think this might be a little confusing, while having `--quiet` is great it would be a little unintuitive that this specific flag does not exec warning, something along the lines of `--skip-warnings` or `--ignore-warnings` would be a bit more straight forward.

Also, when you think of introducing the change, it would not be breaking in the least since you implement new behavior under the new flag. And if we're implementing this under the same flag, people might not even notice the change and that could introduce even more confusion.

Reply:
That's already the case tho for the "quiet" flag - it skips warnings, with that name.

Reply:
> it skips warnings

Well, not quite, correct me if i'm wrong but it[ collects all the results](https://github.com/eslint/eslint/blob/3fc33b0ed008bad509e90f4d98bff4763026171e/lib/cli.js#L429) and filters leaving only errors, thus having lower performance than actually not looking for warnings all-togather.

Reply:
From the user's perspective, though, in almost all cases, they'll only see what is _effectively_ skipped warnings. That it produces the errors and filters them out is an implementation detail.

Reply:
I also feel we can implement this into the `--quiet` flag directly and mark it as a breaking change.

Another alternative that we can consider in case we don't want to do a breaking change is to allow to pass a string value to the `--quite` flag, something like the following:

- `npx eslint . --quiet` - Current behavior.
- `npx eslint . --quiet=errors-only` - Run rules set to errors only.

I'm not sure if the current CLI parser would allow such implementation. Again the naming would be confusing to users.

Reply:
I've modified the document to have this alternative as the main proposal, and included the questions around how it interacts with other flags as an open question. I can see use cases for `--max-warnings` with `--quiet` (eg, not caring about warnings unless there are a lot of them) - so I'm not 100% sure that disabling that flag when `--quiet` is used makes sense, but I can see that allowing it to still work and therefore making warnings run even with quiet mode only in that situation might be extremely confusing